doc/code-principles.md

```diff
@@ -28,3 +28,20 @@ Practically speaking this means:
     comments that merely repeat what the code already says
 -   only have code that actually does something -- don't include any commented
     out code or code that can't be reached or run
+-   minimize the number of return statements in a function
+    -   if there are conditions under which you can bail out easily, it's OK
+        to check for them near the top of the function and return from there
+    -   otherwise, strive to have all returns at or near the end of the
+        function
+
+## Code style/formatting
+
+We rely on automated tools (ESLint and Prettier as of this writing) to keep
+the layout/style of our code uniform for readability. Any new code will be
+required to conform in the commit/PR process. In fact, often the tools will be
+able to adjust the layout of the code for you: if a commit is rejected because
+of formatting non-compliance, try `npm run lint` -- just be aware that it will
+modify your source files. If you just want to see the issues without touching
```

Comment:
For exactly these reasons, I have attempted to make commits fail if there are any changed unstaged files. Please make a trivial change to some file and **don't** add it, and then try to commit. It should fail. If it doesn't, then I need to look into/fix up that bit of the PR. Thanks!

Reply:
A lot of extra stuff has happened in this PR (WebGL, this, ...) -- maybe the description of the PR should change to list those things.  It will help me review, because I'm missing some of them.

Reply:
OK, I have edited the initial description of this PR to include what I hope is a comprehensive list of all of the changes in the PR. It is indeed a lot; my apologies for so much getting stuffed into this PR. It was mostly a consequence of issues that the testing surfaced (and so wouldn't pass in their prior state) but I must sheepishly admit I went a bit overboard with the Histogram hatching and consequent switch to WebGL. The silver lining of my indiscretion is that it surfaced the many difficulties with testing WebGL that would have come up later anyway. Hopefully we can resolve how to handle the documentation-inclusion issue today and things will be in a state that review can continue.

Reply:
Ok, great, I ran through a fake commit cycle and had to lint and add again, and it seems to work well.